src/librustc_target/abi/call/avr.rs

```diff
@@ -0,0 +1,33 @@
+#![allow(non_upper_case_globals)]
+
+use crate::abi::call::{ArgAbi, FnAbi};
+
+fn classify_ret_ty<Ty>(ret: &mut ArgAbi<'_, Ty>) {
+    if ret.layout.is_aggregate() {
+        ret.make_indirect();
+    } else {
+        ret.extend_integer_width_to(8); // Is 8 correct?
+    }
+}
+
+fn classify_arg_ty<Ty>(arg: &mut ArgAbi<'_, Ty>) {
+    if arg.layout.is_aggregate() {
+        arg.make_indirect();
+    } else {
+        arg.extend_integer_width_to(8);
+    }
+}
```

Comment:
You probably want to look at what clang does and cross-reference with a different target that clang treats similarly to AVR, and `rustc_target`'s equivalent code for that target.

I can help with this if you want.

`extend_integer_width_to(8)` won't do anything except to `i1` (i.e. `bool`s), so I don't know if it's relevant.

Reply:
I've done some digging (thanks @eddyb for the tips!) and come to several conclusions and a code update

  * In Clang, there is no explicit AVR argument classification ABI defined. The clang logic falls back to the default  LLVM ABI implementation ([here](https://github.com/llvm/llvm-project/blob/21fee0921d563f407e07b5e28592c2925da3704d/clang/lib/CodeGen/TargetInfo.cpp#L7836))

  * Eddyb is right - the integer width extension to 8 bits can never be hit. I have removed them.

  * There is very little information online about the what the AVR-GCC criterion is for promoting
    byval arguments to passed via an indirect pointer. There is a lot of information about how the
    actual values are laid out in memory and the register file, but very little about the specific logic of   "_these_ kinds of aggregates should be promoted to pointers/pass-by-ref". At the moment though, AVR clang does not match AVR gcc, and that should be fixed first as Rust<->Clang ABI compatibility is first and foremost.

  * Due to the fact that AVR under Clang uses the default, nonspecialized ABI, there will be cases where AVR-Clang generates code that is ABI-incompatible with AVR-GCC for the moment.

I think it is best we update the AVR calling convention in Rust to match the current AVR calling
convention in Clang, even though it isn't currently completely ABI compatible with GCC.

Aside from the redundant width extension, the existing Rust ABI implementation already is compatible with the Clang ABI. A similar example is 32-bit PowerPC in  `librustc_target/abi/call/powerpc.rs`, which implements the same default ABI that AVR Clang uses.

I have raised Clang bug [PR46140](https://bugs.llvm.org/show_bug.cgi?id=46140) to track adding an AVR-GCC compatible ABI implementation to Clang. Once this PR is merged, I can raise a bug on rust-lang/rust to track the AVR-GCC ABI incompatibility for Rust and the subsequent update to `rustc_target` that fixing of Clang PR46140 should trigger.